File: android/src/main/java/com/oney/WebRTCModule/SerializeUtils.java
package com.oney.WebRTCModule;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.ReadableMapKeySetIterator;
import com.facebook.react.bridge.ReadableType;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.WritableMap;

import org.webrtc.*;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class SerializeUtils {
    /**
     * Serialization APIs
     */

    public static ReadableMap serializeVideoCodecInfo(VideoCodecInfo info) {
        WritableMap params = Arguments.createMap();
        params.putString("mimeType", "video/" + info.name);
        return params;
    }

    public static ReadableMap serializeStream(int pcId, String streamReactTag, MediaStream stream) {
        WritableMap params = Arguments.createMap();
        params.putString("streamId", stream.getId());
        params.putString("streamReactTag", streamReactTag);

        WritableArray tracks = Arguments.createArray();

        for (VideoTrack track : stream.videoTracks) {
            tracks.pushMap(SerializeUtils.serializeTrack(pcId, track));
        }
        for (AudioTrack track : stream.audioTracks) {
            tracks.pushMap(SerializeUtils.serializeTrack(pcId, track));
        }

        params.putArray("tracks", tracks);

        return params;
    }

    public static String serializeDirection(RtpTransceiver.RtpTransceiverDirection src) {
        switch (src) {
            case INACTIVE:
                return "inactive";
            case RECV_ONLY:
                return "recvonly";
            case SEND_ONLY:
                return "sendonly";
            case SEND_RECV:
                return "sendrecv";
            case STOPPED:
                return "stopped";
            default:
                throw new Error("Invalid direction");
        }
    }

    public static ReadableMap serializeTrack(int pcId, MediaStreamTrack track) {
        WritableMap trackInfo = Arguments.createMap();
        trackInfo.putString("id", track.id());
        trackInfo.putInt("peerConnectionId", pcId);
        trackInfo.putString("kind", track.kind());
        trackInfo.putBoolean("enabled", track.enabled());
        trackInfo.putString("readyState", track.state().toString().toLowerCase());
        trackInfo.putBoolean("remote", true);
        return trackInfo;
    }

    /**
     * This method is currently missing serializing DtmfSender
     * and transport.
     * TODO: Add transport and dtmf fields to the serialized sender to match the web APIs.
     */
    public static ReadableMap serializeSender(int id, RtpSender sender) {
        WritableMap res = Arguments.createMap();
        res.putString("id", sender.id());
        res.putInt("peerConnectionId", id);
        if (sender.track() != null) {
            res.putMap("track", SerializeUtils.serializeTrack(id, sender.track()));
        }
        res.putMap("rtpParameters", SerializeUtils.serializeRtpParameters(sender.getParameters()));
        return res;
    }

    public static ReadableMap serializeReceiver(int id, RtpReceiver receiver) {
        WritableMap res = Arguments.createMap();
        res.putString("id", receiver.id());
        res.putInt("peerConnectionId", id);
        if (receiver.track() != null) {
            res.putMap("track", SerializeUtils.serializeTrack(id, receiver.track()));
        }
        res.putMap("rtpParameters", SerializeUtils.serializeRtpParameters(receiver.getParameters()));
        return res;
    }

    public static ReadableMap serializeTransceiver(int id, RtpTransceiver transceiver) {
        WritableMap res = Arguments.createMap();
        res.putString("id", transceiver.getSender().id());
        res.putInt("peerConnectionId", id);
        String mid = transceiver.getMid();
        res.putString("mid", mid);
        res.putString("direction", serializeDirection(transceiver.getDirection()));
        RtpTransceiver.RtpTransceiverDirection currentDirection = transceiver.getCurrentDirection();
        if (currentDirection != null) {
            res.putString("currentDirection", SerializeUtils.serializeDirection(transceiver.getCurrentDirection()));
        }
        res.putBoolean("isStopped", transceiver.isStopped());
        res.putMap("receiver", SerializeUtils.serializeReceiver(id, transceiver.getReceiver()));
        res.putMap("sender", SerializeUtils.serializeSender(id, transceiver.getSender()));
        return res;
    }

    public static ReadableMap serializeRtpParameters(RtpParameters params) {
        WritableMap result = Arguments.createMap();
        WritableArray encodings = Arguments.createArray();
        WritableArray codecs = Arguments.createArray();
        WritableArray headerExtensions = Arguments.createArray();
        WritableMap rtcp = Arguments.createMap();

        // Preparing RTCP
        rtcp.putString("cname", params.getRtcp().getCname());
        rtcp.putBoolean("reducedSize", params.getRtcp().getReducedSize());

        // Preparing header extensions
        params.getHeaderExtensions().forEach(extension -> {
            WritableMap extensionMap = Arguments.createMap();
            extensionMap.putInt("id", extension.getId());
            extensionMap.putString("uri", extension.getUri());
            extensionMap.putBoolean("encrypted", extension.getEncrypted());
            headerExtensions.pushMap(extensionMap);
        });

        // Preparing encodings
        params.encodings.forEach(encoding -> {
            WritableMap encodingMap = Arguments.createMap();
            encodingMap.putBoolean("active", encoding.active);
            if (encoding.rid != null) {
                encodingMap.putString("rid", encoding.rid);
            }
            // Since they return integer objects that are nullable,
            // while the map does not accept nullable integer values.
            if (encoding.maxBitrateBps != null) {
                encodingMap.putInt("maxBitrate", encoding.maxBitrateBps);
            }
            if (encoding.maxFramerate != null) {
                encodingMap.putInt("maxFramerate", encoding.maxFramerate);
            }
            if (encoding.scaleResolutionDownBy != null) {
                encodingMap.putDouble("scaleResolutionDownBy", encoding.scaleResolutionDownBy);
            }
            encodings.pushMap(encodingMap);
        });

        // Preparing codecs
        params.codecs.forEach(codec -> {
            WritableMap codecMap = Arguments.createMap();
            codecMap.putInt("payloadType", codec.payloadType);
            codecMap.putString("mimeType", codec.name);
            codecMap.putInt("clockRate", codec.clockRate);
            if (codec.numChannels != null) {
                codecMap.putInt("channels", codec.numChannels);
            }
            // Serializing sdpFmptLine.
            if (!codec.parameters.isEmpty()) {
                final String sdpFmptLineParams = codec.parameters.keySet()
                                                         .stream()
                                                         .map(key -> key + "=" + codec.parameters.get(key))
                                                         .collect(Collectors.joining(";"));
                codecMap.putString("sdpFmtpLine", sdpFmptLineParams);
            }

            codecs.pushMap(codecMap);
        });

        result.putString("transactionId", params.transactionId);
        result.putMap("rtcp", rtcp);
        result.putArray("encodings", encodings);
        result.putArray("codecs", codecs);
        result.putArray("headerExtensions", headerExtensions);
        if (params.degradationPreference != null) {
            result.putString("degradationPreference", params.degradationPreference.toString());
        }

        return result;
    }

    /**
     * Parsing APIs
     */

    public static RtpParameters updateRtpParameters(ReadableMap updateParams, RtpParameters rtpParams) {
        // Preparing encodings
        ReadableArray encodingsArray = updateParams.getArray("encodings");
        List<RtpParameters.Encoding> encodings = rtpParams.encodings;
        if (encodingsArray.size() != encodings.size()) {
            return null;
        }

        for (int i = 0; i < encodingsArray.size(); i++) {
            ReadableMap encodingUpdate = encodingsArray.getMap(i);
            RtpParameters.Encoding encoding = encodings.get(i);
            // Dealing with nullable Integers
            Integer maxBitrate = encodingUpdate.hasKey("maxBitrate") ? encodingUpdate.getInt("maxBitrate") : null;
            Integer maxFramerate = encodingUpdate.hasKey("maxFramerate") ? encodingUpdate.getInt("maxFramerate") : null;
            Double scaleResolutionDownBy = encodingUpdate.hasKey("scaleResolutionDownBy")
                    ? encodingUpdate.getDouble("scaleResolutionDownBy")
                    : null;

            encoding.active = encodingUpdate.getBoolean("active");
            encoding.rid = encodingUpdate.getString("rid");
            encoding.maxBitrateBps = maxBitrate;
            encoding.maxFramerate = maxFramerate;
            encoding.scaleResolutionDownBy = scaleResolutionDownBy;
        }

        if (updateParams.hasKey("degradationPreference")) {
            rtpParams.degradationPreference =
                    RtpParameters.DegradationPreference.valueOf(updateParams.getString("degradationPreference"));
        }

        return rtpParams;
    }

    public static MediaStreamTrack.MediaType parseMediaType(String type) {
        switch (type) {
            case "audio":
                return MediaStreamTrack.MediaType.MEDIA_TYPE_AUDIO;
            case "video":
                return MediaStreamTrack.MediaType.MEDIA_TYPE_VIDEO;
            default:
                throw new Error("Unknown media type");
        }
    }

    public static RtpTransceiver.RtpTransceiverDirection parseDirection(String src) {
        switch (src) {
            case "sendrecv":
                return RtpTransceiver.RtpTransceiverDirection.SEND_RECV;
            case "sendonly":
                return RtpTransceiver.RtpTransceiverDirection.SEND_ONLY;
            case "recvonly":
                return RtpTransceiver.RtpTransceiverDirection.RECV_ONLY;
            case "inactive":
                return RtpTransceiver.RtpTransceiverDirection.INACTIVE;
                // Here we ignore the "stopped" direction because user code should
                // never set it.
        }
        throw new Error("Invalid direction");
    }

    private static RtpParameters.Encoding parseEncoding(ReadableMap params) {
        RtpParameters.Encoding encoding = new RtpParameters.Encoding(params.getString("rid"), true, 1.0);

        if (params.hasKey("active")) {
            encoding.active = params.getBoolean("active");
        }
        if (params.hasKey("maxBitrate")) {
            encoding.maxBitrateBps = params.getInt("maxBitrate");
        }
        if (params.hasKey("maxFramerate")) {
            encoding.maxFramerate = params.getInt("maxFramerate");
        }
        if (params.hasKey("scaleResolutionDownBy")) {
            encoding.scaleResolutionDownBy = params.getDouble("scaleResolutionDownBy");
        }

        return encoding;
    }

    public static RtpTransceiver.RtpTransceiverInit parseTransceiverOptions(ReadableMap map) {
        if (map == null) {
            return null;
        }

        RtpTransceiver.RtpTransceiverDirection direction = RtpTransceiver.RtpTransceiverDirection.SEND_RECV;
        List<String> streamIds = new ArrayList<>();
        List<RtpParameters.Encoding> sendEncodings = new ArrayList<>();

        if (map.hasKey("direction")) {
            String directionRaw = map.getString("direction");
            if (directionRaw != null) {
                direction = SerializeUtils.parseDirection(directionRaw);
            }
        }
        if (map.hasKey("streamIds")) {
            ReadableArray rawStreamIds = map.getArray("streamIds");
            if (rawStreamIds != null) {
                for (int i = 0; i < rawStreamIds.size(); i++) {
                    streamIds.add(rawStreamIds.getString(i));
                }
            }
        }
        if (map.hasKey("sendEncodings")) {
            ReadableArray encodingsArray = map.getArray("sendEncodings");
            if (encodingsArray != null) {
                for (int i = 0; i < encodingsArray.size(); i++) {
                    ReadableMap encoding = encodingsArray.getMap(i);
                    sendEncodings.add(SerializeUtils.parseEncoding(encoding));
                }
            }
        }

        return new RtpTransceiver.RtpTransceiverInit(direction, streamIds, sendEncodings);
    }
}

File: ios/RCTWebRTC/SerializeUtils.m
#import "SerializeUtils.h"

@implementation SerializeUtils
+ (NSDictionary *)transceiverToJSONWithPeerConnectionId:(NSNumber *)id
                                            transceiver:(RTCRtpTransceiver *_Nonnull)transceiver {
    NSMutableDictionary *result = [NSMutableDictionary new];

    result[@"id"] = transceiver.sender.senderId;
    result[@"peerConnectionId"] = id;
    result[@"mid"] = transceiver.mid;
    result[@"direction"] = [SerializeUtils serializeDirection:transceiver.direction];

    RTCRtpTransceiverDirection currentDirection;
    if ([transceiver currentDirection:&currentDirection]) {
        result[@"currentDirection"] = [SerializeUtils serializeDirection:currentDirection];
    }

    result[@"isStopped"] = [NSNumber numberWithBool:transceiver.isStopped];
    result[@"receiver"] = [SerializeUtils receiverToJSONWithPeerConnectionId:id receiver:transceiver.receiver];
    result[@"sender"] = [SerializeUtils senderToJSONWithPeerConnectionId:id sender:transceiver.sender];

    return result;
}

+ (NSMutableArray *)constructTransceiversInfoArrayWithPeerConnection:(RTCPeerConnection *)peerConnection {
    NSMutableArray *transceiverUpdates = [NSMutableArray new];

    for (RTCRtpTransceiver *transceiver in peerConnection.transceivers) {
        NSMutableDictionary *transceiverUpdate = [NSMutableDictionary new];

        RTCRtpTransceiverDirection currentDirection;
        BOOL hasCurrentDirection = [transceiver currentDirection:&currentDirection];
        if (hasCurrentDirection) {
            NSString *currentDirectionSerialized = [SerializeUtils serializeDirection:currentDirection];
            transceiverUpdate[@"currentDirection"] = currentDirectionSerialized;
        }

        transceiverUpdate[@"transceiverId"] = transceiver.sender.senderId;
        transceiverUpdate[@"mid"] = transceiver.mid;
        transceiverUpdate[@"isStopped"] = [NSNumber numberWithBool:transceiver.isStopped];
        transceiverUpdate[@"senderRtpParameters"] = [SerializeUtils parametersToJSON:transceiver.sender.parameters];
        transceiverUpdate[@"receiverRtpParameters"] = [SerializeUtils parametersToJSON:transceiver.receiver.parameters];

        [transceiverUpdates addObject:transceiverUpdate];
    }

    return transceiverUpdates;
}

+ (NSDictionary *)senderToJSONWithPeerConnectionId:(NSNumber *)id sender:(RTCRtpSender *)sender {
    NSMutableDictionary *senderDictionary = [NSMutableDictionary new];
    senderDictionary[@"id"] = sender.senderId;
    senderDictionary[@"peerConnectionId"] = id;

    if (sender.track) {
        senderDictionary[@"track"] = [SerializeUtils trackToJSONWithPeerConnectionId:id track:sender.track];
    }

    senderDictionary[@"rtpParameters"] = [SerializeUtils parametersToJSON:sender.parameters];

    return senderDictionary;
}

+ (NSDictionary *)receiverToJSONWithPeerConnectionId:(NSNumber *)id receiver:(RTCRtpReceiver *)receiver {
    NSMutableDictionary *receiverDictionary = [NSMutableDictionary new];
    receiverDictionary[@"id"] = receiver.receiverId;
    receiverDictionary[@"peerConnectionId"] = id;

    if (receiver.track) {
        receiverDictionary[@"track"] = [SerializeUtils trackToJSONWithPeerConnectionId:id track:receiver.track];
    }

    receiverDictionary[@"rtpParameters"] = [SerializeUtils parametersToJSON:receiver.parameters];

    return receiverDictionary;
}

+ (NSDictionary *)parametersToJSON:(RTCRtpParameters *)params {
    NSMutableDictionary *paramsDictionary = [NSMutableDictionary new];

    NSMutableDictionary *rtcpDictionary = [NSMutableDictionary new];
    rtcpDictionary[@"cname"] = params.rtcp.cname;
    rtcpDictionary[@"reducedSize"] = [NSNumber numberWithBool:params.rtcp.isReducedSize];

    NSMutableArray *headerExtensions = [NSMutableArray new];

    for (RTCRtpHeaderExtension *extension in params.headerExtensions) {
        NSMutableDictionary *extensionDictionary = [NSMutableDictionary new];
        extensionDictionary[@"id"] = [NSNumber numberWithInt:extension.id];
        extensionDictionary[@"uri"] = extension.uri;
        extensionDictionary[@"encrypted"] = [NSNumber numberWithBool:extension.isEncrypted];

        [headerExtensions addObject:extensionDictionary];
    }

    NSMutableArray *encodings = [NSMutableArray new];

    for (RTCRtpEncodingParameters *encoding in params.encodings) {
        NSMutableDictionary *encodingDictionary = [NSMutableDictionary new];

        encodingDictionary[@"active"] = [NSNumber numberWithBool:encoding.isActive];

        if (encoding.rid) {
            encodingDictionary[@"rid"] = encoding.rid;
        }
        if (encoding.maxBitrateBps) {
            encodingDictionary[@"maxBitrate"] = encoding.maxBitrateBps;
        }
        if (encoding.maxFramerate) {
            encodingDictionary[@"maxFramerate"] = encoding.maxFramerate;
        }
        if (encoding.scaleResolutionDownBy) {
            encodingDictionary[@"scaleResolutionDownBy"] = encoding.scaleResolutionDownBy;
        }

        [encodings addObject:encodingDictionary];
    }

    NSMutableArray *codecs = [NSMutableArray new];

    for (RTCRtpCodecParameters *codec in params.codecs) {
        NSMutableDictionary *codecDictionary = [NSMutableDictionary new];

        codecDictionary[@"payloadType"] = [NSNumber numberWithInt:codec.payloadType];
        codecDictionary[@"mimeType"] = codec.name;
        codecDictionary[@"clockRate"] = codec.clockRate;

        if (codec.numChannels) {
            codecDictionary[@"channels"] = codec.numChannels;
        }

        if (codec.parameters.count) {
            NSMutableArray *parts = [NSMutableArray arrayWithCapacity:codec.parameters.count];
            [codec.parameters
                enumerateKeysAndObjectsUsingBlock:^(id _Nonnull key, id _Nonnull value, BOOL *_Nonnull stop) {
                    [parts addObject:[NSString stringWithFormat:@"%@=%@", key, value]];
                }];

            codecDictionary[@"sdpFmtpLine"] = [parts componentsJoinedByString:@";"];
        }

        [codecs addObject:codecDictionary];
    }

    paramsDictionary[@"transactionId"] = params.transactionId;
    paramsDictionary[@"rtcp"] = rtcpDictionary;
    paramsDictionary[@"headerExtensions"] = headerExtensions;
    paramsDictionary[@"encodings"] = encodings;
    paramsDictionary[@"codecs"] = codecs;

    if (params.degradationPreference) {
        paramsDictionary[@"degradationPreference"] = params.degradationPreference;
    }

    return paramsDictionary;
}

+ (NSDictionary *)trackToJSONWithPeerConnectionId:(NSNumber *)id track:(RTCMediaStreamTrack *)track {
    NSString *readyState;
    switch (track.readyState) {
        case RTCMediaStreamTrackStateLive:
            readyState = @"live";
            break;
        case RTCMediaStreamTrackStateEnded:
            readyState = @"ended";
            break;
    }

    return @{
        @"id" : track.trackId,
        @"peerConnectionId" : id,
        @"kind" : track.kind,
        @"enabled" : [NSNumber numberWithBool:track.isEnabled],
        @"readyState" : readyState,
        @"remote" : [NSNumber numberWithBool:YES],
    };
}

+ (NSString *)serializeDirection:(RTCRtpTransceiverDirection)direction {
    if (direction == RTCRtpTransceiverDirectionInactive) {
        return @"inactive";
    } else if (direction == RTCRtpTransceiverDirectionRecvOnly) {
        return @"recvonly";
    } else if (direction == RTCRtpTransceiverDirectionSendOnly) {
        return @"sendonly";
    } else if (direction == RTCRtpTransceiverDirectionSendRecv) {
        return @"sendrecv";
    } else if (direction == RTCRtpTransceiverDirectionStopped) {
        return @"stopped";
    }
    return nil;
}

+ (RTCRtpTransceiverDirection)parseDirection:(NSString *)direction {
    if ([direction isEqual:@"inactive"]) {
        return RTCRtpTransceiverDirectionInactive;
    } else if ([direction isEqual:@"recvonly"]) {
        return RTCRtpTransceiverDirectionRecvOnly;
    } else if ([direction isEqual:@"sendonly"]) {
        return RTCRtpTransceiverDirectionSendOnly;
    } else if ([direction isEqual:@"sendrecv"]) {
        return RTCRtpTransceiverDirectionSendRecv;
    } else if ([direction isEqual:@"stopped"]) {
        return RTCRtpTransceiverDirectionStopped;
    }

    return RTCRtpTransceiverDirectionInactive;
}

+ (RTCRtpEncodingParameters *)parseEncoding:(NSDictionary *)params {
    RTCRtpEncodingParameters *encoding = [RTCRtpEncodingParameters new];

    if (params[@"rid"] != nil) {
        [encoding setRid:params[@"rid"]];
    }
    if (params[@"active"] != nil) {
        [encoding setIsActive:((NSNumber *)params[@"active"]).boolValue];
    }
    if (params[@"maxBitrate"] != nil) {
        [encoding setMaxBitrateBps:(NSNumber *)params[@"maxBitrate"]];
    }
    if (params[@"maxFramerate"] != nil) {
        [encoding setMaxFramerate:(NSNumber *)params[@"maxFramerate"]];
    }
    if (params[@"scaleResolutionDownBy"] != nil) {
        [encoding setScaleResolutionDownBy:(NSNumber *)params[@"scaleResolutionDownBy"]];
    }

    return encoding;
}

+ (RTCRtpTransceiverInit *)parseTransceiverOptions:(NSDictionary *)params {
    RTCRtpTransceiverInit *transceiverInit = [RTCRtpTransceiverInit new];

    NSString *direction = [params objectForKey:@"direction"];
    if (direction) {
        [transceiverInit setDirection:[SerializeUtils parseDirection:direction]];
    }

    NSArray *streamIds = [params objectForKey:@"streamIds"];
    if (streamIds) {
        [transceiverInit setStreamIds:streamIds];
    }

    NSArray *encodingsArray = [params objectForKey:@"sendEncodings"];
    if (encodingsArray) {
        NSMutableArray<RTCRtpEncodingParameters *> *sendEncodings = [NSMutableArray new];
        for (NSDictionary *encoding in encodingsArray) {
            [sendEncodings addObject:[self parseEncoding:encoding]];
        }
        [transceiverInit setSendEncodings:sendEncodings];
    }

    return transceiverInit;
}

+ (NSDictionary *)streamToJSONWithPeerConnectionId:(NSNumber *)id
                                            stream:(RTCMediaStream *)stream
                                    streamReactTag:(NSString *)streamReactTag {
    NSMutableDictionary *streamDictionary = [NSMutableDictionary new];

    streamDictionary[@"streamId"] = stream.streamId;
    streamDictionary[@"streamReactTag"] = streamReactTag;

    NSMutableArray *tracks = [NSMutableArray new];

    for (RTCAudioTrack *audioTrack in stream.audioTracks) {
        [tracks addObject:[SerializeUtils trackToJSONWithPeerConnectionId:id track:audioTrack]];
    }

    for (RTCVideoTrack *videoTrack in stream.videoTracks) {
        [tracks addObject:[SerializeUtils trackToJSONWithPeerConnectionId:id track:videoTrack]];
    }

    streamDictionary[@"tracks"] = tracks;

    return streamDictionary;
}
@end
